### ISSUE TITLE
fleet.service not starting on virtual box Non Vagrant Deis installation

### ISSUE THREAD

**Issue by pdhar-tibco** [deis/deis#3546]
I am trying to install deis on 3 coreos instances running on my mac OS X virtualbox vm host. I am not using Vagrant , instead installed coreos independently. I have configured the 3 vm's to use 2 network interfaces each(public+private=local3.deisapp.com) and I got etcd running properly. When I proceeded to run the "deisctl install platform" step it just hangs there. previously I tried to install deis on amazon ec2 cluster and i was able to install. I would like to know if I missed something.

The outputs in the logs show
---------------
 journalctl -e -u fleet.service
Apr 21 22:59:18 localhost systemd[1]: Started fleet daemon.
Apr 21 22:59:18 localhost systemd[1]: fleet.service: main process exited, code=exited, status=203/EXEC
Apr 21 22:59:18 localhost systemd[1]: Unit fleet.service entered failed state.
Apr 21 22:59:18 localhost systemd[1]: fleet.service failed.
-----------------
systemctl status fleet.service
● fleet.service - fleet daemon
   Loaded: loaded (/usr/lib64/systemd/system/fleet.service; static; vendor preset: disabled)
  Drop-In: /run/systemd/system/fleet.service.d
           └─20-cloudinit.conf
        /etc/systemd/system/fleet.service.d
           └─99-upgrade-fleet-091.conf
   Active: activating (auto-restart) (Result: exit-code) since Tue 2015-04-21 22:58:37 UTC; 9s ago
  Process: 2571 ExecStart=/run/deis/bin/fleetd-0.9.2 (code=exited, status=203/EXEC)
 Main PID: 2571 (code=exited, status=203/EXEC)
Apr 21 22:58:37 localhost systemd[1]: Unit fleet.service entered failed state.
Apr 21 22:58:37 localhost systemd[1]: fleet.service failed.
-------
The fleet configuration at /run/systemd/system/fleet.service.d has the following contents
localhost fleet.service.d # cat 20-cloudinit.conf 
[Service]
Environment="FLEET_ETCD_REQUEST_TIMEOUT=3"
Environment="FLEET_PUBLIC_IP=172.17.8.100"
and 
 cat /etc/systemd/system/fleet.service.d/99-upgrade-fleet-091.conf                   
[Service]
ExecStart=
ExecStart=/run/deis/bin/fleetd-0.9.2
---------------ifconfig -------
enp0s3: flags=4163<UP,BROADCAST,RUNNING,MULTICAST>  mtu 1500
        inet 10.108.0.83  netmask 255.255.254.0  broadcast 10.108.1.255
        inet6 fe80::a00:27ff:fe2b:eed3  prefixlen 64  scopeid 0x20<link>
        ether 08:00:27:2b:ee:d3  txqueuelen 1000  (Ethernet)
        RX packets 24036  bytes 3801667 (3.6 MiB)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 4457  bytes 1412057 (1.3 MiB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0

enp0s8: flags=4163<UP,BROADCAST,RUNNING,MULTICAST>  mtu 1500
        inet 172.17.8.100  netmask 255.255.255.0  broadcast 172.17.8.255
        inet6 fe80::a00:27ff:fe2c:92e4  prefixlen 64  scopeid 0x20<link>
        ether 08:00:27:2c:92:e4  txqueuelen 1000  (Ethernet)
        RX packets 17  bytes 6850 (6.6 KiB)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 35  bytes 5224 (5.1 KiB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0

lo: flags=73<UP,LOOPBACK,RUNNING>  mtu 65536
        inet 127.0.0.1  netmask 255.0.0.0
        inet6 ::1  prefixlen 128  scopeid 0x10<host>
        loop  txqueuelen 0  (Local Loopback)
        RX packets 2623  bytes 1092037 (1.0 MiB)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 2623  bytes 1092037 (1.0 MiB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0
----- etcd.service logs ----
Apr 21 22:12:10 localhost etcd[561]: [etcd] Apr 21 22:12:10.307 INFO      | etcd server [name ed488a160da74d16a6a7a04d0709b1d1, listen on :4001, advertised url http://172.17.8.100:4001]
Apr 21 22:12:10 localhost etcd[561]: [etcd] Apr 21 22:12:10.307 INFO      | peer server [name ed488a160da74d16a6a7a04d0709b1d1, listen on :7001, advertised url http://172.17.8.100:7001]
Apr 21 22:12:10 localhost etcd[561]: [etcd] Apr 21 22:12:10.307 INFO      | ed488a160da74d16a6a7a04d0709b1d1 starting in peer mode
Apr 21 22:12:10 localhost etcd[561]: [etcd] Apr 21 22:12:10.307 INFO      | ed488a160da74d16a6a7a04d0709b1d1: state changed from 'initialized' to 'follower'.
Apr 21 22:12:10 localhost etcd[561]: [etcd] Apr 21 22:12:10.309 INFO      | failed to get local etcd versions: open /usr/libexec/etcd/internal_versions/: no such file or directory
Apr 21 22:12:13 localhost etcd[561]: [etcd] Apr 21 22:12:13.309 INFO      | ed488a160da74d16a6a7a04d0709b1d1: snapshot of 20007 events at index 20007 completed
Apr 21 22:12:13 localhost etcd[561]: [etcd] Apr 21 22:12:13.379 INFO      | ed488a160da74d16a6a7a04d0709b1d1: state changed from 'follower' to 'candidate'.
Apr 21 22:12:13 localhost etcd[561]: [etcd] Apr 21 22:12:13.380 INFO      | ed488a160da74d16a6a7a04d0709b1d1: state changed from 'candidate' to 'leader'.
Apr 21 22:12:13 localhost etcd[561]: [etcd] Apr 21 22:12:13.380 INFO      | ed488a160da74d16a6a7a04d0709b1d1: leader changed from '' to 'ed488a160da74d16a6a7a04d0709b1d1'.
Apr 21 22:26:22 localhost etcd[561]: [etcd] Apr 21 22:26:22.697 INFO      | ed488a160da74d16a6a7a04d0709b1d1: snapshot of 10002 events at index 30009 completed

**Comment by pdhar-tibco** [comment]
I noticed that the fleet was having version 0.9.1 and there was an upgrade script /etc/systemd/system/update-fleet-091.service which is run when systemctl start fleet is called.
When I checked if the /run/deis/bin/fleet-0.9.2 was downloaded as per the script it wasn't
x 1 root root   44 Apr 21 03:49 99-upgrade-fleet-091.conf -> /run/deis/conf/fleetd-092-custom-binary.conf
localhost fleet.service.d # cd /run/deis/
localhost deis # ls
bin  conf
localhost deis # cd bin
localhost bin # ls
get_image  preseed
localhost bin # systemctl start fleet
localhost bin # ls
get_image  preseed
localhost bin # systemctl status fleet
● fleet.service - fleet daemon
   Loaded: loaded (/usr/lib64/systemd/system/fleet.service; static; vendor preset: disabled)
  Drop-In: /run/systemd/system/fleet.service.d
           └─20-cloudinit.conf
        /etc/systemd/system/fleet.service.d
           └─99-upgrade-fleet-091.conf
   Active: activating (auto-restart) (Result: exit-code) since Wed 2015-04-22 00:01:57 UTC; 3s ago
  Process: 657 ExecStart=/run/deis/bin/fleetd-0.9.2 (code=exited, status=203/EXEC)
 Main PID: 657 (code=exited, status=203/EXEC)

Apr 22 00:01:57 localhost systemd[1]: Unit fleet.service entered failed state.
Apr 22 00:01:57 localhost systemd[1]: fleet.service failed.
localhost bin # ls -al
total 8
drwxr-xr-x 2 root root  80 Apr 22 00:00 .
drwxr-xr-x 4 root root  80 Apr 22 00:00 ..
-rwxr-xr-x 1 root root 473 Apr 22 00:00 get_image
-rwxr-xr-x 1 root root 300 Apr 22 00:00 preseed
localhost bin # pwd
/run/deis/bin
localhost bin # 

When the upgrade part of the upgrade-fleet-091.service was run from command line
/usr/bin/bash -c 'if fleetd --version | grep -q 0.9.1; then curl -sSL --retry 5 --retry-delay 2 -o /run/deis/bin/fleetd-0.9.2 https://s3-us-west-2.amazonaws.com/opdemand/fleetd-v0.9.2 && chmod +x /run/deis/bin/fleetd-0.9.2 && mkdir -p /etc/systemd/system/fleet.service.d/ && ln -s /run/deis/conf/fleetd-092-custom-binary.conf /etc/systemd/system/fleet.service.d/99-upgrade-fleet-091.conf; else rm -f /etc/systemd/system/fleet.service.d/99-upgrade-fleet-091.conf; fi'

the /run/deis/bin has the fleet-0.9.2 in it
localhost system # ls /run/deis/bin/
fleetd-0.9.2  get_image  preseed

I wonder if the script is actually getting executed when "systemctl start fleet" is run and is meeting the pre-conditions in the script
**-------**
[Unit]
Description=Upgrade fleet if system fleet is v0.9.1, which has a known bug
Before=fleet.service
ConditionPathIsSymbolicLink=!/etc/systemd/system/fleet.service.d/99-upgrade-fleet-091.conf
drwxr-xr-x 2 root root 4096 Apr 21 03:49 .
drwxr-xr-x 6 root root 4096 Apr 22 00:05 ..
lrwxrwxrwx 1 root root   44 Apr 21 03:49 99-upgrade-fleet-091.conf -> /run/deis/conf/fleetd-092-custom-binary.conf
**--- PREV--**
systemctl status fleet
● fleet.service - fleet daemon
   Loaded: loaded (/usr/lib64/systemd/system/fleet.service; static; vendor preset: disabled)
  Drop-In: /run/systemd/system/fleet.service.d
           └─20-cloudinit.conf
        /etc/systemd/system/fleet.service.d
           └─99-upgrade-fleet-091.conf
   Active: activating (auto-restart) (Result: exit-code) since Wed 2015-04-22 00:01:57 UTC; 3s ago
  Process: 657 ExecStart=/run/deis/bin/fleetd-0.9.2 (code=exited, status=203/EXEC)
 Main PID: 657 (code=exited, status=203/EXEC)

**---- AFTER manual exec ----**
Apr 22 00:06:03 localhost fleetd-0.9.2[751]: INFO reconcile.go:332: AgentReconciler completed task: type=LoadUnit job=deis-store-daemon.service reason="unit scheduled here but not loaded"
Apr 22 00:06:03 localhost fleetd-0.9.2[751]: INFO reconcile.go:332: AgentReconciler completed task: type=LoadUnit job=deis-store-metadata.service reason="unit scheduled her... not loaded"
Apr 22 00:06:03 localhost fleetd-0.9.2[751]: INFO reconcile.go:332: AgentReconciler completed task: type=LoadUnit job=deis-store-monitor.service reason="unit scheduled here... not loaded"
Apr 22 00:06:03 localhost fleetd-0.9.2[751]: INFO reconcile.go:332: AgentReconciler completed task: type=LoadUnit job=deis-store-volume.service reason="unit scheduled here but not loaded"
Apr 22 00:06:03 localhost fleetd-0.9.2[751]: INFO reconcile.go:332: AgentReconciler completed task: type=ReloadUnitFiles job=N/A reason="always reload unit files"
Apr 22 00:06:04 localhost fleetd-0.9.2[751]: INFO manager.go:248: Writing systemd unit deis-store-gateway@1.service (957b)
Apr 22 00:06:04 localhost fleetd-0.9.2[751]: INFO manager.go:184: Instructing systemd to reload units
Apr 22 00:06:04 localhost fleetd-0.9.2[751]: INFO reconcile.go:332: AgentReconciler completed task: type=LoadUnit job=deis-store-gateway@1.service reason="unit scheduled he... not loaded"
Apr 22 00:06:04 localhost fleetd-0.9.2[751]: INFO reconcile.go:332: AgentReconciler completed task: type=ReloadUnitFiles job=N/A reason="always reload unit files"
Apr 22 00:09:51 localhost systemd[1]: Started fleet daemon.
Hint: Some lines were ellipsized, use -l to show in full.

------------
I am not sure why the fleetd-0.9.2 is not getting downloaded when run by the start script

**Comment by pdhar-tibco** [comment]
The condition was never met as it was a link and the condition check for ! a link

**Comment by carmstrong** [comment]
Thanks @pdhar-tibco - this is #3547. Is this a fresh provision or did you see this happen on a restart? Let's try to figure this out in #3547.

**Comment by pdhar-tibco** [comment]
This was a fresh provision as of yesterday. I started using the same user-dat.example that I used for my ec2 evaluation with new discovery-url for the virtual box.
The changes I had to make to make it this far is
1. created 2nd network interface using HostOnlyAdapter (vboxnet2 =172.17.8.0/24)
2. cloned 2 vm’s from first one.
3. saw 2 different ip’s on each vm console
4. ssh ed to one and saw etcd not running.
5. added /etc/environment 
cat /etc/environment
COREOS_PUBLIC_IPV4=10.108.0.85
COREOS_PRIVATE_IPV4=172.17.8.102
6. found fleetd not running and preventing deis installation
7. ran the /etc/systemd/system/upgrade-fleet-091.services manually to download the fleetd-0.9.2
8. now running deis install platform 
pranabs-mbp:~ pdhar$ deisctl install platform
● ▴ ■
■ ● ▴ Installing Deis...
▴ ■ ●

Storage subsystem...
deis-store-daemon.service: loaded                                 
deis-store-gateway@1.service: loaded                                 
deis-store-monitor.service: loaded                                 
deis-store-volume.service: loaded                                 
deis-store-metadata.service: loaded                                 
Logging subsystem...
deis-logger.service: loaded                                 
deis-logspout.service: loaded                                 
Control plane...
deis-registry@1.service: loaded                                 
deis-database.service: loaded                                 
deis-builder.service: loaded                                 
deis-controller.service: loaded                                 
Data plane...
deis-publisher.service: loaded                                 
Routing mesh...
deis-router@1.service: loaded
